test/langtools/tools/javac/processing/model/util/types/TestAnnotationStripping.java

```diff
@@ -0,0 +1,196 @@
+
+                System.err.println();
+                System.err.println();
+            }
+
+            if (failures > 0)
+                throw new RuntimeException(failures + " failures occured.");
+        }
+        return true;
+    }
+
+    void testVacuous(TypeMirror tm ) {
+        try {
+            var result = vacuousTypes.stripAnnotations(tm);
+            messager.printError("Unexpected non-exceptional result returned" +  result);
+        } catch(UnsupportedOperationException uoe) {
+            ; // Expected
+        }
+    }
+
+    private int failures = 0;
+
+    void checkExpectedTypeAnnotations(AnnotatedConstruct ac, TypeMirror expectedAnnotation) {
+        List<? extends AnnotationMirror> annotations = ac.getAnnotationMirrors();
+        if (annotations.size() != 1) {
+            failures++;
+            System.err.println("\t\t\tUnexpected annotations size: " + annotations.size());
+        } else if (!typeUtils.isSameType(annotations.get(0).getAnnotationType(), expectedAnnotation)) {
+            failures++;
+            System.err.println("\t\t\tUnexpected annotations type: " + annotations);
+        }
+    }
+
+    void checkEmptyAnnotations(AnnotatedConstruct ac) {
+        System.err.println("\t" + ac);
+        if (ac == null)
+            return;
+        else {
+            List<? extends AnnotationMirror> annotations = ac.getAnnotationMirrors();
+            int count = 0;
+            for (AnnotationMirror annotation : annotations) {
+              if (((TypeElement) annotation.getAnnotationType().asElement()).getQualifiedName().contentEquals("jdk.internal.ValueBased")) {
+                continue;
+              }
+              count++;
+            }
+            if (count != 0) {
+                failures++;
+                System.err.println(ac.getClass());
+                System.err.println("\t\t\tUnexpected nonzero annotations size: " + annotations);
+            }
+        }
+    }
+
+    void checkEqualTypeAndAnnotations(TypeMirror tm1, TypeMirror tm2) {
+        if (!typeUtils.isSameType(tm1, tm2)) {
+            failures++;
+            System.err.printf("Unequal types %s and %s.%n", tm1, tm2);
+        }
+
+        if (!Objects.equals(tm1.getAnnotationMirrors(), tm1.getAnnotationMirrors())) {
+            failures++;
+            System.err.printf("Unequal annotations on and %s.%n", tm1, tm2);
+        }
+    }
+
+}
+
+/*
+ * Class to host annotations for testing
+ */
+class HostClass {
+    public static @TestTypeAnnotation("foo") Integer foo() {return null;}
+
+    public static @TestTypeAnnotation("foo2") int foo2() {return 0;}
+
+    public static @TestTypeAnnotation("foo3") String foo3() {return null;}
+
+    public static  java.util.@TestTypeAnnotation("foo4")Set foo4() {return null;}
+
+    public static  String @TestTypeAnnotation("foo5")[]  foo5() {return null;}
+
+    public static  java.util. @TestTypeAnnotation("foo6") Set < @TestTypeAnnotation("foo7") String> foo6() {return null;}
```

Comment:
Consider adding a case for type variable usages:

```
public static <@TestTypeAnnotation("foo8") T extends @TestTypeAnnotation("foo9") String> @TestTypeAnnotation("foo10") T foo8() {return null;}
```

This one raises a couple of questions about type annotations associated with the declaration of the type variable. I think probably only the annotations on the use of the type var need to be stripped, not the declaration.

Currently this causes failures for the assertion that `asElement()` has no annotations, because the annotations are only stripped from the type corresponding to the use of the type variable, not from the corresponding declaration.

Similarly `TypeVariable#getUpperBound` would report annotations.